### ISSUE TITLE
[BUG] [CNL] IPC timeout during suspend-resume tests

### ISSUE THREAD

**Issue by lyakh** [thesofproject/linux#2498]
I've been getting an "IPC timed out" error during my weekly Mantis testing:
```
[ 7339.996692] [93112] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x80010000
[ 7340.010861] [93112] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x20140000 successful
[ 7340.010874] [93112] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40010000
[ 7340.514720] sof-audio-pci 0000:00:1f.3: error: ipc timed out for 0x40010000 size 76
[ 7340.514724] sof-audio-pci 0000:00:1f.3: info: preventing DSP entering D3 state to preserve context
[ 7340.514732] [93112] sof-audio-pci 0000:00:1f.3: unknown ROM status value ffffffff
[ 7340.514775] sof-audio-pci 0000:00:1f.3: invalid header size 0xffffffff. FW oops is bogus
[ 7340.514779] sof-audio-pci 0000:00:1f.3: error: unexpected fault 0xffffffff trace 0xffffffff
[ 7340.514789] sof-audio-pci 0000:00:1f.3: error: hda irq intsts 0x00000000 intlctl 0xc0000000 rirb 00
[ 7340.514792] sof-audio-pci 0000:00:1f.3: error: dsp irq ppsts 0x00000000 adspis 0x00000000
[ 7340.514799] sof-audio-pci 0000:00:1f.3: error: host status 0x00000000 dsp status 0x00000000 mask 0x00000003
[ 7340.514803] sof-audio-pci 0000:00:1f.3: ctx_save ipc error -110, proceeding with suspend
[ 7340.514906] [93112] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1010f0f successful
[ 7340.514913] [93112] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0xf0f successful
[ 7340.514917] [93112] sof-audio-pci 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 7340.514967] [93112] sof-audio-pci 0000:00:1f.3: Debug PCIR: 00000010 at  00000044
[ 7340.515009] [93112] sof-audio-pci 0000:00:1f.3: Current DSP power state: D3
```
We've had similar issues with different codes or on different platforms: #2334 https://github.com/thesofproject/sof/issues/3541 and it's unclear whether they all have the same cause or not. This specific one triggers during "check-suspend-resume-with-audio.sh -m capture" on "HDA Analog"

**Comment by xiulipan** [comment]
@lyakh I just tried to reproduce this test on some CML platforms, no luck with 3 tries. Will do more test

**Comment by lyakh** [comment]
> @lyakh I just tried to reproduce this test on some CML platforms, no luck with 3 tries. Will do more test

@xiulipan you mean 3 single tests or 3 loops of N tests? For me it typically takes between 10 and 15 repetitions.

**Comment by xiulipan** [comment]
@lyakh Thanks for the clarify, I will try more time.

**Comment by wwittbrx** [comment]
@lyakh What configs are you using when building the FW (cannonlake_defconfig, cannonlake_chrome_overridedefconfig)?

Also, what commits are you building for FW? I can reproduce similar issue on slim driver when running f97e9f0 with Chrome config. No reproduction on e5d17f3.

**Comment by lyakh** [comment]
> @lyakh What configs are you using when building the FW (cannonlake_defconfig, cannonlake_chrome_overridedefconfig)?
> 
> Also, what commits are you building for FW? I can reproduce similar issue on slim driver when running f97e9f0 with Chrome config. No reproduction on e5d17f3.

@wwittbrx I'm using cannonlake_defconfig. The newest commit with which I was able to reproduce the problem was 80ef3cf95e4954b6177b9171c4931f253a0808d2 Even easier it seems to be to reproduce it with 1b17974b3bda4a2723246612e294c226e6dea88a

**Comment by lyakh** [comment]
Update:
1. I have no problem reproducing the bug with a clean build, based on 1b17974b3bda4a2723246612e294c226e6dea88a
2. I've only once seen the bug with a build, based on 80ef3cf95e4954b6177b9171c4931f253a0808d2 and that build wasn't clean, i.e. the build directory contained files from previous builds.
3. A large number of attempts to reproduce the bug with a clean build, based on 80ef3cf95e4954b6177b9171c4931f253a0808d2 failed.
4. This seems to indicate, that something between those two commits fixed the problem. But there are only two commits between those two hashes and they are irrelevant for CNL defconfig. I.e. those two hashes should really behave the same.

**Comment by keyonjie** [comment]
@lyakh since @xiulipan can't reproduce with the same recipe, can you help to debug it more once you can reproduce it stablly.

**Comment by lyakh** [comment]
Triggered again in a kmod-load-unload test as documented in #2535 which is closed as duplicate of this issue.

**Comment by lyakh** [comment]
update: the easiest way currently to reproduce this problem on my CNL hardware is to run `sudo test-case/check-kmod-load-unload.sh -l 5` from https://github.com/thesofproject/sof-test, in my tests the error always triggers within those 5 loops. However, after removing the line 
https://github.com/thesofproject/sof/blob/9b5ba319387aaaf8cd606dd6054464c015b7ce2c/src/ipc/handler.c#L580 I was able to run 15 loops of the test with no error.

**Comment by lyakh** [comment]
update 2: as of last Friday's weekly test the problem didn't occur. Also wasn't able to reproduce it manually earlier during last week. How about I check again end of this week, and if the bug doesn't show up again, we close this bug (until it re-appears)?

**Comment by mengdonglin** [comment]
@lyakh Okay. Thank you!

**Comment by lyakh** [comment]
Last Friday same result: no more reproduction, closing this bug as agreed

**Comment by mengdonglin** [comment]
@lyakh  Thank you for double check!

**Comment by lyakh** [comment]
I think it is the same issue, but maybe it is not. We've had several bugs filed with different IPC fields reading 0xffffffff or 0xff000000 or something similar, this is one of them. Now I came across a failure of https://github.com/thesofproject/sof/pull/3642 on Mantis (again) with no complete dmesg or firmware logs, but the test quotes a single kernel log line:
> [  747.300322] kernel: sof-audio-pci 0000:00:1f.3: error: DSP trace buffer overflow 4294967295 bytes. Total messages -1

which means a trace position IPC has been read with `.overflow` and `.messages` set to 0xffffffff. Checking the code this cannot be done by the firmware, so, this leads me to think, that it's again mailbox content flipping to 0xffffffff possibly again while suspending the DSP.

**Comment by plbossart** [comment]
@lyakh The Mantis issue also happened for one of my SoundWire PRs, see https://sof-ci.01.org/linuxpr/PR2601/build4953/devicetest/?model=CML_MANTIS_HDA&testcase=check-kmod-load-unload


````
sh-cml-mantis-hda-01 kernel: [  715.595077] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40020000: GLB_PM_MSG: CTX_RESTORE
sh-cml-mantis-hda-01 kernel: [  715.595194] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
sh-cml-mantis-hda-01 kernel: [  715.595196] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
sh-cml-mantis-hda-01 kernel: [  715.595228] sof-audio-pci 0000:00:1f.3: error: DSP trace buffer overflow 4294967295 bytes. Total messages -1
sh-cml-mantis-hda-01 kernel: [  715.595293] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
sh-cml-mantis-hda-01 kernel: [  715.595345] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
````

It doesn't look random any longer, I can't recall having see two reports on the same day? This also happens smack while we are doing PM-related IPC...

**Comment by plbossart** [comment]
@lyakh if this isn't a known issue any longer please close. We can always reopen a new issue later.

**Comment by lyakh** [comment]
Closing for a lack of recent activity or evidence
